chainer/datasets/text_dataset.py

```diff
@@ -0,0 +1,172 @@
+            the number of files. Arguments are lines loaded from each file.
+            The filter function must return True to accept the line, or
+            return False to skip the line.
+
+    """
+
+    _fps = None
+
+    def __init__(
+            self, paths, encoding=None, errors=None, newline=None,
+            filter_func=None):
+        if isinstance(paths, six.string_types):
+            paths = [paths]
+        elif len(paths) == 0:
+            raise ValueError('at least one text file must be specified')
+
+        if isinstance(encoding, six.string_types) or encoding is None:
+            encoding = [encoding] * len(paths)
+        if isinstance(errors, six.string_types) or errors is None:
+            errors = [errors] * len(paths)
+        if isinstance(newline, six.string_types) or newline is None:
+            newline = [newline] * len(paths)
+
+        if not (len(paths) == len(encoding) == len(errors) == len(newline)):
+            raise ValueError(
+                'length of each option must match with the number of '
+                'text files to read')
+
+        self._paths = paths
+        self._encoding = encoding
+        self._errors = errors
+        self._newline = newline
+        self._fps = None
+
+        self._open()
+
+        # Line number is 0-origin.
+        # `lines` is a list of line numbers not filtered; if no filter_func is
+        # given, it is range(linenum)).
+        # `bounds` is a list of cursor positions of line boundaries for each
+        # file, i.e. i-th line of k-th file starts at `bounds[k][i]`.
+        linenum = 0
+        lines = []
+        bounds = tuple([[0] for _ in self._fps])
+        while True:
+            data = [fp.readline() for fp in self._fps]
+            if not all(data):  # any of files reached EOF
+                if any(data):  # not all files reached EOF
+                    raise ValueError(
+                        'number of lines in files does not match')
+                break
+            for i, fp in enumerate(self._fps):
+                bounds[i].append(fp.tell())
+            if filter_func is not None and filter_func(*data):
+                lines.append(linenum)
+            linenum += 1
+
+        if filter_func is None:
+            lines = six.moves.range(linenum)
+
+        self._bounds = bounds
+        self._lines = lines
+        self._lock = threading.Lock()
+
+    def __del__(self):
+        self._close()
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['_fps']
+        del state['_lock']
+        return state
+
+    def __setstate__(self, state):
+        self._close()
+        self.__dict__ = state
+        self._open()
+        self._lock = threading.Lock()
+
+    def __len__(self):
+        return len(self._lines)
+
+    def _open(self):
+        if self._fps is not None:
+            return
+        self._fps = [
+            io.open(
```

Comment:
Do you need to close fps when an error occurs?

Reply:
Unlike `open` in Python 2.7, we don't have to care about errors because `io.open` returns wrapper class that implements `__del__` that calls `close`. https://docs.python.org/2/library/io.html#io.IOBase.__del__
I removed redundant close code.